src/pysight/post_processing/flim.py

```diff
@@ -1,205 +1,111 @@
-        """
-        lowest = hist.argmin()
-        if lowest < 2:
-            raise ValueError
-        return lowest
-
-    def _fit_decay(self, y):
-        """
-        Use a curve fitting method to find the parameters of the
-        exponential decay of the data
-        """
-        xdata = np.arange(len(y)) / (self.data.config["binwidth"] * 10 ** 9)
-        popt, pcov = scipy.optimize.curve_fit(
-            self.single_exp, xdata, y, (y.max(), 4, np.median(y))
-        )
-        return popt, pcov
-
-    @staticmethod
-    def single_exp(x, amp, tau, c):
-        """
-        A single decaying exponential fit, to be used with
-        SciPy's ``curve_fit`` function.
-        :param np.ndarray x: "Time" vector
-        :param Union[float,int] amp: Starting amplitude
-        :param Union[float,int] tau: Expected lifetime
-        :param Union[float,int] c: Baseline level of the function
-        """
-        return amp * np.exp(-x / tau) + c
-
-
-@attr.s(frozen=True)
-class ExpFitParams:
-    """
-    Returned parameters from fitting a decay curve to a
-    single exponential decay function.
+    # clean up
+    data.drop(columns=['bin_of_dim0'])
+    return data
 
-    Parameters:
-        :param float amplitude: Initial amplitude of decay.
-        :param float tau: Fluorophore lifetime.
-        :param float c: Constant offset of the decay above baseline.
-    """
 
-    amplitude = attr.ib(validator=instance_of(float))
-    tau = attr.ib(validator=instance_of(float))
-    c = attr.ib(validator=instance_of(float))
+def calc_lifetime(data, bins_bet_pulses=125) -> float:
+    """
+    calculate lifetime by creating an histogramming the photons by time of arrival,
+    we would then expect to see an exponential decay to which we can fit an exponential curve to estimate the lifetime.
+    a Savitzky Golay filter is applied to the histogram to smooth out noise.
+    :param data: array of photon times of arrival
+    :param bins_bet_pulses: number of bins to split the photons into
+    :return: estimate lifetime (int)
+    """
+    hist, edges = np.histogram(data, bins_bet_pulses)
+    if len(data) < 20 or hist.max() < 5:
+        # photon count is too small
+        return np.nan
+
+    hist = np.roll(hist, -hist.argmax())
+    sg_hist = savgol_filter(hist, 9, 2)
+
+    # find bin with least amount of photons
+    lowest_bin_idx = hist.argmin()
+    if lowest_bin_idx < 5:
+        return np.nan
+    sg_hist = sg_hist[:lowest_bin_idx]
+
+    # estimate tau
+    try:
+        popt, _ = curve_fit(_exp_decay, np.arange(len(sg_hist)), sg_hist,
+                            p0=[sg_hist.max(), 1 / np.average(sg_hist), sg_hist.min()],
+                            maxfev=1_000,
+                            bounds=(0, [len(data), 70, np.inf]))
+        tau = 1/popt[1]
+    except (RuntimeError, ValueError):
+        tau = np.nan
+
+    if not (70.0 > tau > 0.0001):
+        return np.nan
+
+    return tau/10  # divide by 10 to fix units
```

Comment:
You should divide by the Multiscaler's sampling rate. This division should occur outside of this file - frame chunk should divide the whole array at once after the Tau calculation business is over.

Reply:
Is the Multiscaler's sampling rate what's referred to as Multiscaler binwidth in the config?

If so, I understand that it is given in units of seconds. How would I go about realizing the units of the tau estimated from `time_rel_line` and `abs_time`? Is it always going to be the same?

Reply:
The units of `time_rel_line` and `abs_time` are indeed always the sames - they're always a pure number representing the number of time intervals since the start of the experiment. We want to show the Tau with units of ns, so the conversion is quite straightforward given that you know what the binwidth of the multiscaler is.

Reply:
So,  we want this to be: `tau = tau/multiscaler_binwidth*1e-9`?